modules/rostests/kmtests/rtl/RtlStrSafe.c

```diff
@@ -0,0 +1,195 @@
+    
+    /* STATUS_BUFFER_OVERFLOW tests */
+
+    UsString.Buffer = BufferSmall;
+    UsString.Length = 0;
+    UsString.MaximumLength = 2;
+    
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_BUFFER_OVERFLOW);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+    ok_eq_wchar(UsString.Buffer[0], L'A');
+    ok_eq_wchar(UsString.Buffer[1], L'\0');
+    for (i = 2; i < sizeof(BufferSmall); i++)
+        ok_eq_wchar(UsString.Buffer[i], 0xAAAA);
+
+    UsString.Buffer = BufferSmall;
+    UsString.Length = 0;
+    UsString.MaximumLength = 7;
+
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"0123", L"4567", L"89AB"), STATUS_BUFFER_OVERFLOW);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+    ok_eq_wchar(UsString.Buffer[0], L'0');
+    ok_eq_wchar(UsString.Buffer[1], L'1');
+    ok_eq_wchar(UsString.Buffer[2], L'2');
+    ok_eq_wchar(UsString.Buffer[3], L'3');
+    ok_eq_wchar(UsString.Buffer[4], L' ');
+    ok_eq_wchar(UsString.Buffer[5], L'4');
+    ok_eq_wchar(UsString.Buffer[6], L'\0');
+    for (i = 7; i < sizeof(BufferSmall); i++)
+        ok_eq_wchar(UsString.Buffer[i], 0xAAAA);
+
+    ///* STATUS_INVALID_PARAMETER tests */
+
+    ok_eq_hex(RtlUnicodeStringPrintf(NULL, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, NULL, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+
+    UsStringNull.Buffer = NULL;
+    UsStringNull.Length = 0;
+    UsStringNull.MaximumLength = 0;
+    ok_eq_bool(RtlUnicodeStringPrintf(&UsStringNull, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+
+    /* Test  for buffer overruns */
+
+    RtlFillMemory(Buffer, sizeof(Buffer), 0xAA);
+    RtlFillMemory(OvrBuffer, sizeof(OvrBuffer), 0xAA);
+    UsString.Buffer = Buffer;
+    UsString.Length = 0;
+    UsString.MaximumLength = 16 * sizeof(WCHAR);
+
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"abc", L"def", L"ghi"), STATUS_SUCCESS);
+    ok_eq_uint(UsString.Length, sizeof(L"abc def ghi") - sizeof(WCHAR));
+    ok_eq_wchar(UsString.Buffer[11], L'\0');
+    ok_eq_uint(0, memcmp(OvrBuffer + 12, Buffer + 12, sizeof(Buffer) - (12 * sizeof(WCHAR))));
+
+    // None of these functions should have crashed.
+    KmtEndSeh(STATUS_SUCCESS);
+}
+
+TESTAPI
+Test_RtlUnicodeStringPrintfEx()
+{
+    WCHAR Buffer[32];
+    WCHAR BufferSmall[16];
+    WCHAR OvrBuffer[1024];
+    UNICODE_STRING UsString, RemString;
+    const WCHAR FormatStringInts[] = L"%d %d %d";
+    const WCHAR FormatStringStrs[] = L"%s %s %s";
+    const WCHAR Result[] = L"1 2 3";
+    int i;
+
+    /* No zeros (Don't assume UNICODE_STRINGS are NULL terminated) */
+
+    RtlFillMemory(Buffer, sizeof(Buffer), 0xAA);
+    RtlFillMemory(BufferSmall, sizeof(BufferSmall), 0xAA);
+
+    UsString.Buffer = Buffer;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(Buffer);
```

Comment:
RtlInitEmptyUnicodeString comes to mind.

Reply:
Fixed that. Thanks!